import PyObject from 'components/PyObject';
import { DynamicMetaTags } from 'components/MetaTags';

<DynamicMetaTags
  title="Conditional Execution/Branching | Dagster"
  description="Dagster is a system for building modern data applications."
/>

# Conditional Execution/Branching

## What are optional outputs?

An output can be marked as `is_optional` in an <PyObject module="dagster" object="OutputDefinition" displayText="OutputDefinitions" />.
This means that the output does not necessarily have to be yielded by the solid.

If an optional output is not yielded, all downstream solids that depend
on the output will simply skip, such as short-circuiting or branching.

For example:

```python literalinclude caption=conditional_execution.py
file:/dagster_examples/how_tos/conditional_execution.py
lines:8-11
```

## How do I conditionally branch in my pipeline?

By taking advantage of optional outputs and skips, we can implement
branching in our pipeline.

For example, let's define a solid that has two optional outputs, and
only yields one based on a conditional:

```python literalinclude caption=conditional_execution.py
file:/dagster_examples/how_tos/conditional_execution.py
lines:14-24
```

Then, we can have two downstream solids `path_1` and `path_2` that
depend on output "a" and output "b", respectively:

```python literalinclude caption=conditional_execution.py
file:/dagster_examples/how_tos/conditional_execution.py
lines:27-34
```

Finally, we can wire these solids up in a pipeline:

```python literalinclude caption=conditional_execution.py
file:/dagster_examples/how_tos/conditional_execution.py
lines:37-41
```

Depending on the value of the conditional, only `path_1` or `path_2`
will execute.

## How do I manage error handling in a solid?

There are two types of possible errors that can happen in the body of a
solid's `compute_fn`:

1.  A python runtime exception that is unexpected
2.  Running into a known exception or invalid state

_Case 1: Python runtime exception_

When an exception is thrown in a solid's `compute_fn`, the error is
caught by a user error boundary and surfaced in the event stream. Any
downstream solids that depend on the solid will be marked as skipped and
will not execute.

_Case 2: Known invalid state or unrecoverable failure_

If you run into a known failure state, you can `raise` a <PyObject module="dagster" object="Failure" /> event
from a solid to in order to notify the Dagster executor of the failure as well
as return structured metadata about the failure.

For example, a scenario where you want to "fail successfully" is when
you have an external data store that is not ready/available.

Any downstream solids that depend on the solid will be marked as skipped
and will not execute.

```python literalinclude caption=hello_cereal.py
file:/dagster_examples/how_tos/conditional_execution.py
lines:52-59
```
